react/features/face-centering/middleware.js

```diff
@@ -0,0 +1,135 @@
+import { FACE_BOX_EVENT_TYPE } from './constants';
+
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const state = getState();
+    const { faceCoordinatesSharing } = state['features/base/config'];
+    const isInTileView = shouldDisplayTileView(state);
+
+    if (!getCurrentConference(getState())) {
+        return next(action);
+    }
+
+    if (action.type === CONFERENCE_JOINED) {
+        if (faceCoordinatesSharing?.enabled) {
+            dispatch(loadWorker());
+
+            if (isInTileView) {
+                dispatch(startFaceRecognition());
+            }
+        }
+
+        // allow using remote face centering data when local face centering is not enabled
+        action.conference.on(
+            JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
+            (...args) => {
+                if (args && args.length >= 2) {
+                    const [ { _id }, eventData ] = args;
+
+                    if (eventData.type === FACE_BOX_EVENT_TYPE) {
+                        dispatch({
+                            type: UPDATE_FACE_COORDINATES,
+                            faceBox: eventData.faceBox,
+                            id: _id
+                        });
+                    }
+                }
+            });
+
+        return next(action);
+    }
+
+    if (!faceCoordinatesSharing?.enabled) {
+        return next(action);
+    }
+
+    // might need to change logic here if object-fit: cover will be used outside of tileview in the future
+    if (action.type === SET_TILE_VIEW) {
+        if (action.enabled) {
+            dispatch(startFaceRecognition());
+        } else {
+            dispatch(stopFaceRecognition());
+        }
+
+        return next(action);
+    }
+
+    if (!isInTileView) {
+        return next(action);
+    }
+
+    switch (action.type) {
+    case CONFERENCE_WILL_LEAVE : {
+        dispatch(stopFaceRecognition());
+
+        return next(action);
+    }
+    case TRACK_UPDATED: {
+        const { videoType, type } = action.track.jitsiTrack;
+
+        if (videoType === 'camera') {
+            const { muted, videoStarted } = action.track;
+
+            if (videoStarted === true) {
+                dispatch(startFaceRecognition());
+            }
+            if (muted !== undefined) {
+                if (muted) {
+                    dispatch(stopFaceRecognition());
+                } else {
+                    dispatch(startFaceRecognition());
+                    type === 'presenter' && changeTrack(action.track);
+                }
+            }
+        }
+
+        return next(action);
+    }
+    case TRACK_ADDED: {
+        const { mediaType, videoType } = action.track;
+
+        if (mediaType === 'presenter' && videoType === 'camera') {
+            dispatch(startFaceRecognition());
+            changeTrack(action.track);
+        }
+
+        return next(action);
+    }
+    case TRACK_REMOVED: {
+        const { videoType } = action.track.jitsiTrack;
```

Comment:
Shouldn't we be checking for the `TRACK_REMOVED`, `TRACK_ADDED` and `TRACK_UPDATED` if the change is related to the local tracks?

Reply:
Seems like the 'camera' value for videoType is only set for local tracks.

Reply:
Not really.

The whole story is that we treat undefined as camera in order to be able to reduce the size of the signaling and skip sending it initially until it changes. So if you switch to desktop sharing and then back to camera the video type will be camera.

But you can pretty easy check for the `local` prop in the track from the store.

Reply:
Maybe it will be good to check if the track is a video track as well. Just in case that we have some strange cases for the local tracks where the video type is undefined.